### ISSUE TITLE
Frame representation is not respected when instanced using a representation

### ISSUE THREAD

**Issue by astrojuanlu** [astropy/astropy#6435]
Yes, the title of this issue is awful... Please feel free to edit for clarity.

When I instance a frame using a subclass of `BaseRepresentation` as the container of the data (instead of passing the data directly to the initializer of the frame), I would expect the representation of the frame to be the one I used, but it's not:

```python
In [3]: GCRS(CartesianRepresentation(*iss.r))
Out[3]: 
<GCRS Coordinate (obstime=J2000.000, obsgeoloc=( 0.,  0.,  0.) m, obsgeovel=( 0.,  0.,  0.) m / s): (ra, dec, distance) in (deg, deg, km)
    ( 281.73052421,  51.41284322,  6774.76995296)>

In [4]: _.representation
Out[4]: astropy.coordinates.representation.SphericalRepresentation
```

Is there a reason for this? I'm aware that there are other ways to instance a frame though, which behave differently:

```python
In [5]: GCRS(x=iss.r[0], y=iss.r[1], z=iss.r[2], representation="cartesian")
Out[5]: 
<GCRS Coordinate (obstime=J2000.000, obsgeoloc=( 0.,  0.,  0.) m, obsgeovel=( 0.,  0.,  0.) m / s): (x, y, z) in km
    ( 859.07256, -4137.20368,  5295.56871)>

In [6]: GCRS(x=iss.r[0], y=iss.r[1], z=iss.r[2], representation=CartesianRepresentation)
Out[6]: 
<GCRS Coordinate (obstime=J2000.000, obsgeoloc=( 0.,  0.,  0.) m, obsgeovel=( 0.,  0.,  0.) m / s): (x, y, z) in km
    ( 859.07256, -4137.20368,  5295.56871)>
```

**Comment by Juanlu001** [comment]
Also, specifying the representation twice works as well:

```python
In [18]: GCRS(CartesianRepresentation(*iss.r), representation="cartesian")
Out[18]: 
<GCRS Coordinate (obstime=J2000.000, obsgeoloc=( 0.,  0.,  0.) m, obsgeovel=( 0.,  0.,  0.) m / s): (x, y, z) in km
    ( 859.07256, -4137.20368,  5295.56871)>
```

**Comment by mhvk** [comment]
Yes, this is bizarre and should be changed! Labeling it as a bug.

**Comment by Juanlu001** [comment]
This was closed in https://github.com/astropy/astropy/pull/6440, so closing!

**Comment by astrojuanlu** [comment]
#6440 fixed #6436, but this issue is still present:

```
>>> r1 = CartesianRepresentation(0, 1, 0)
>>> icrs = ICRS(r1)
>>> icrs
<ICRS Coordinate: (ra, dec, distance) in (deg, deg, )
    (90., 0., 1.)>
>>> icrs.representation_type
<class 'astropy.coordinates.representation.SphericalRepresentation'>
```

(discovered while investigating #10253)

_@astrojuanlu stares at his past self_

**Comment by astrojuanlu** [comment]
While debugging this I found an interesting behavior:

```
>>> from astropy.coordinates import ICRS
>>> from astropy import units as u
>>> icrs = ICRS(ra=37.4*u.deg, dec=-55.8*u.deg)
>>> icrs.representation_type
<class 'astropy.coordinates.representation.SphericalRepresentation'>
>>> icrs.data
<UnitSphericalRepresentation (lon, lat) in deg
    (37.4, -55.8)>
```

so, even if the data gets converted to a `UnitSphericalRepresentation` object, the representation type ends up being `SphericalRepresentation`.

Wondering if this is another instance of [this subtlety stated by @eteq](https://github.com/astropy/astropy/issues/7784#issuecomment-432775315):

> this touches on the subtle distinction that data can be any representation type, but representation_type is sort of the "view" into that data.

**Comment by mhvk** [comment]
@astrojuanlu - the difference is here between what is shown in the `repr` and what is actually used:
```
from astropy.coordinates import ICRS, CartesianRepresentation
r1 = CartesianRepresentation(0, 1, 0)
icrs = ICRS(r1)
icrs.data
# This is what is actually used by code internally
# <CartesianRepresentation (x, y, z) [dimensionless]
    (0., 1., 0.)>
icrs
# But the repr has a "preferred" representation.
<ICRS Coordinate: (ra, dec, distance) in (deg, deg, )
    (90., 0., 1.)>
```

This relates to a long-standing irritation with the code deciding for me that degrees are better than hours:
```
ICRS(ra='12h13m14s', dec='+45d')
Out[9]: 
<ICRS Coordinate: (ra, dec) in deg
    (183.30833333, 45.)>
```
See issue #6184 - the conclusion there was that for backwards compatibility we cannot easily change the default units that are shown, but we can at least give users the option.

Arguably, though, if one passes in a representation, the user has made a much stronger statement about what the preference is, so to me it seems that the fact that the code converts it to spherical just for `repr` is a bug.

**Comment by astrojuanlu** [comment]
Thanks @mhvk - I think I noticed a slightly different issue though: when doing `icrs = ICRS(ra=37.4*u.deg, dec=-55.8*u.deg)` the representation class is _inferred_, however `icrs.representation_type` and the class of `icrs.data` end up being different, which is weird.

> Arguably, though, if one passes in a representation, the user has made a much stronger statement about what the preference is

Totally agree with this.

**Comment by mhvk** [comment]
Hmm, the difference between `representation_type` and the `type(icrs.data)` indeed seems a separate issue; I think the `get_representation_cls` code is too focussed on doing the right thing for initialization --
https://github.com/astropy/astropy/blob/40bd0e487a65f92fd23eeb73738d57cf8a2232ad/astropy/coordinates/baseframe.py#L719-L741

**Comment by mhvk** [comment]
Actually, checking further, `representation_type` seems to be used just for `repr` - setting it changes what is shown and what attributes the frame has, but it does not affect `.data`. That said, a simple fix for the `repr` issue would be that, if a representation is passed in, the code sets the `self._representation` private attribute.

**Comment by ayshih** [comment]
> Arguably, though, if one passes in a representation, the user has made a much stronger statement about what the preference is, so to me it seems that the fact that the code converts it to spherical just for `repr` is a bug.

While I don't necessarily disagree with this sentiment, I'll point out that some of the code in SunPy "relies" on the current behavior.  Nearly all transformations are easier to compute in Cartesian, so I construct those output frames by passing in the transformed Cartesian representation, and then expect the default representation type for the frame to dictate how the data actually gets shown to the user in the `repr`.  I believe – but haven't actually verified – that this results in better under-the-hood performance with multiple transformations.

It's admittedly odd that `.data` can be a different representation than what shows in the `repr`, but what I tell SunPy users is to never use `.data`.  If they want to access the component data and the type of representation matters, I tell them to be explicit: e.g., use `.spherical`.

**Comment by astrojuanlu** [comment]
> so I construct those output frames by passing in the transformed Cartesian representation

@ayshih If I understand correctly, changing this would not affect you then? The only difference is that passing a `CartesianRepresentation` would set the representation type to cartesian, instead of leaving the default of the frame. What representation is used for `repr` should not affect calculations (as this only has an effect for REPL users) and the underlying `.data` does not affect you either (in the places where you follow the advice of not relying on it, and using `.spherical` or `.cartesian` instead).

**Comment by ayshih** [comment]
If I took no action on SunPy side, you are correct that there would be no calculation impacts, but there would still be impacts:

- Because `representation_type` dictates what components are directly accessible as attributes (e.g., `.lon` versus `.x`), this definitely affects more than just REPL users.
- The `repr` for coordinates returned from transformations would show Cartesian components instead of the default components, which would be undesirable in our documentation and elsewhere.

How to fix SunPy would depend on the implementation of such a change in Astropy.  If I just need to explicitly set the desired `representation_type` so that it is not inferred from the representation of `.data`, that should be fine.  However, if the Astropy implementation actively *prohibits* me from specifying a `representation_type` that is not the same as the representation type of `.data`, then the only way I can achieve the same outward behavior would be to first convert the representation to the desired type before passing it in.  I'd like to avoid the performance impacts and precision impacts of unnecessary conversions.

**Comment by astrojuanlu** [comment]
> If I just need to explicitly set the desired `representation_type` so that it is not inferred from the representation of `.data`, that should be fine.

I think this should be the way to go. In order of precedence:

1. setting `representation_type=` in the frame initializer should _always_ take precedence,
2. if not given, but a explicit representation class with data was used, that should be picked as `representation_type` (**this is the change we would need to implement**, the one I suggested originally when I opened this issue, and would close #10253 automatically),

and then there's a third point:

3. if a representation class is inferred from the `*args`, that should also be picked as `representation_type`.

This change is more debatable: on one hand, I think it would be good for consistency (see my [earlier comment here](https://github.com/astropy/astropy/issues/6435#issuecomment-688265124) about a "slightly different issue")), but on the other hand it renders the "default representation" of the frames kind of useless.

I would be fine if we do (1) and (2) only.

**Comment by mhvk** [comment]
@astrojuanlu - I agree with your analysis, and think I'd include (3) as well - if the user puts in x,y,z, it would seem logical that the default would be cartesian.

@ayshih - your point about attributes being influenced by a change like this is well-taken. So, just to be sure, you are fine with a change like @astrojuanlu proposed, as long as you can still pass in `representation_type` to ensure the default repr/attributes are different from what is input, correct? I.e., `ICRS(CartesianRepresentation(x, y, z), representation_type='spherical')` would set `.data` to a cartesian representation (as now), but `repr` and attributes would be spherical.

On option 3, would we also want `ICRS(x=x, y=y, z=z, representation_type='spherical')` to be possible? I ask since I think I'd prefer to call that too ambiguous and raise on it (feeling that in the input, `representation_type` makes more sense to tell which representation and thus which named arguments you are using for input, having an exception only in the case that the input is a representation already). But an ideal is not worth pursuing if it breaks existing code! And it may also make sense to just do 1 & 2.

p.s. Just so we are clear about what would be needed if we were even stricter, and forced `representation_type` to always match the input: in that case, one could set it after initialization and nothing new would be calculated. But I do think `Frame(representation, representation_type=OtherRepresentation)` is unambiguous, so do not see we have to go there.

**Comment by ayshih** [comment]
> So, just to be sure, you are fine with a change like @astrojuanlu proposed, as long as you can still pass in `representation_type` to ensure the default repr/attributes are different from what is input, correct?

Yup, in principle.  In practice, it's a bit more tricky: the transformed frames are currently instantiated via `.realize_frame()`, which has no way to pass in the `representation_type`.  Unless that is changed, I'll need to either switch to the private `._replicate()` or change the `representation_type` after instantiation.

**Comment by mhvk** [comment]
Good to notice that now rather than later. Given that all the other methods allow one to pass in `**kwargs`, which get passed to the initializer, the simplest might be to do the same here...

These examples do worry me a bit - there is a bit of a mix between telling how to display and how to interpret input in the current scheme.

**Comment by astrojuanlu** [comment]
> Yup, in principle. In practice, it's a bit more tricky: the transformed frames are currently instantiated via `.realize_frame()`, which has no way to pass in the `representation_type`. Unless that is changed, I'll need to either switch to the private `._replicate()` or change the `representation_type` after instantiation.

I was using `._replicate()` in poliastro for this exact reason, which I reported in #7784. Seems like we're converging :slightly_smiling_face: Let us tackle that one first, so we can pave the way to the changes we propose here, if we all agree.

> These examples do worry me a bit - there is a bit of a mix between telling how to display and how to interpret input in the current scheme.

I agree, and in fact after going over #4903 and #6184 again to try to gather some context, I think there are three aspects to consider:

1. How to interpret input in the frame initializer (i.e. what `**kwargs` to accept, using `representation_type` and the corresponding keys of `.representation_component_names`)
2. How to store frame `._data` (i.e. how the frame attributes are transformed to representation attributes using the `.representation_component_names` mapping)
3. How to display frame `._data` (i.e. how the representation attributes are again transformed to frame attributes using the `.representation_component_names` inverse mapping in ._data_repr()`)

I'll scratch my head a little bit more and see if I can find a solution to this long standing annoyance that is not too disruptive with current code.
